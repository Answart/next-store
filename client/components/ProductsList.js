import PropTypes from 'prop-types';
import Link from 'next/link';
import { StyledProductsList, StyledProductsListItem } from './styles/ProductStyles';
import PriceTag from './PriceTag';
import { DeleteProduct } from './Buttons';


const ProductsList = ({ products, editView, userId }) => (
  <StyledProductsList>
    {products.map(prdct => {
      const viewerIsCreator = !!userId
        ? prdct.user.id === userId
        : false;
      const variant = prdct.variants.length
        ? prdct.variants[0]
        : null;
      return (
        <StyledProductsListItem key={prdct.id} >
          <Link href={{
            pathname: "/buy",
            query: { id: prdct.id }
          }}><a>
            {prdct.image && (
              <img width="400" height="610" src={prdct.image.image_url} alt={prdct.title} />
            )}
          </a></Link>

          <div className="prdct-itm-info">
            <Link href={{
              pathname: "/buy",
              query: { id: prdct.id }
            }}><a className='prdct-itm-title'>
              {prdct.title}
            </a></Link>

            {variant && (
              <div className="prdct-itm-price">
                <PriceTag
                  price={variant.price}
                  sale={variant.sale}
                  salePrice={variant.salePrice}
                />
              </div>
            )}

            {prdct.online && variant && (
              <div className="prdct-itm-avail">
                {variant.availability}
              </div>
            )}
          </div>

          {viewerIsCreator && editView && (
            <span className="prdct-itm-actns">
              <Link href={{
                pathname: "/product/edit",
                query: { id: prdct.id }
              }}><a>
                Edit
              </a></Link>

              <DeleteProduct className="undrln-btn"
                id={prdct.id}
                userName={prdct.user.name}
              >Delete</DeleteProduct>
            </span>
          )}

          {!prdct.online && (
            <span><i className="prdct-itm-actns">
              (Offline)
            </i></span>
          )}
        </StyledProductsListItem>
      )
    })}
  </StyledProductsList>
);

ProductsList.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    online: PropTypes.bool.isRequired,
    user: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    image: PropTypes.shape({
      image_url: PropTypes.string.isRequired
    }).isRequired,
    variants: PropTypes.arrayOf(PropTypes.shape({
      price: PropTypes.number.isRequired,
      sale: PropTypes.bool.isRequired,
      salePrice: PropTypes.number.isRequired,
      availability: PropTypes.string.isRequired
    })).isRequired
  })).isRequired,
  userId: PropTypes.string,
  editView: PropTypes.bool.isRequired
};


export default ProductsList;
