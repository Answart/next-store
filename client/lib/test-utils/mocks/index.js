export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrderItem, mockOrder,
} from './typeDefs';
export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
  createImageMutationMock, createImageMutationErrorMock,
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock,
} from './resolvers';
