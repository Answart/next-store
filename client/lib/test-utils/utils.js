import casual from 'casual';


casual.seed(777);

const fakeUser = () => ({
  __typename: 'User',
  id: 'us3r1d',
  name: casual.name,
  email: casual.email,
  password: 'unique-password',
  permissions: ['ADMIN'],
  cart: [],
});

const fakeImage = () => ({
  __typename: 'Image',
  id: '1m4g31d',
  cloudinary_id: "111111",
  name: "peggswatch1",
  width: 22,
  height: 22,
  transformation: "",
  image_url: "peggswatch1.jpg",
  large_image_url: "peggswatch2.jpg",
});

const fakeProduct = () => ({
  __typename: 'Product',
  id: 'pr0duct1d',
  department: "accessories",
  title: "Peggs Gold Edition Analog Watch Peggs Gold Edition Analog ",
  description: "Limited Edition watch from the 2018 Fall fashion line.",
  category: "jewelry",
  brand: "Peggs",
  online: false,
  user: fakeUser(),
  variants: [],
  image: fakeImage(),
});

const fakeVariant = () => ({
  __typename: 'Variant',
  id: 'v4r13nt1d',
  availability: "4 in Stock!",
  color: "white",
  price: 35,
  quantity: 4,
  sale: true,
  salePrice: 30,
  size: "S",
  image: fakeImage(),
  product: fakeProduct()
});

const fakeCartItem = overrides => ({
  __typename: 'CartItem',
  id: 'c4rt1t3m1d',
  quantity: 3,
  user: fakeUser(),
  variant: fakeVariant(),
  ...overrides,
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: '0rd3r1d',
  charge: 'ch_123',
  total: 4000,
  status: 'processing',
  quantity: 3,
  subtotal: 20,
  shipping: 1.5,
  tax: 1.02,
  total: 20,
  sales_tax_rate: 1.02,
  shipping_rate: 1.02,
  payment: 'some payment',
  items: [],
  buyer: fakeUser(),
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  updatedAt: '2018-04 - 06T19: 24: 16.000Z',
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: '0rd3r1t3m1d',
  title: 'order item title',
  price: 12,
  quantity: 2,
  color: 'green',
  size: 'L',
  image_url: 'some_img_url',
  seller: fakeUser(),
  variant: fakeVariant(),
  order: fakeOrder(),
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  updatedAt: '2018-04 - 06T19: 24: 16.000Z',
});


export {
  fakeUser,
  fakeImage,
  fakeProduct,
  fakeVariant,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
