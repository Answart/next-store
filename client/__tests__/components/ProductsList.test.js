import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import ProductsList from '../../components/ProductsList';
import {
  mockUser, mockProducts
} from '../../lib/test-utils/mocks';


describe('<ProductsList />', () => {
  let wrapper;
  const variantAction = jest.fn();
  beforeAll(() => {
    wrapper = mount(
      <MockedProvider mocks={[]}>
        <ProductsList
          products={mockProducts}
          editView={true}
          userId={mockUser.id}
        />
      </MockedProvider>
    );
  });
  afterAll(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });

  it('matches the snap shot', async () => {
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('ProductsList'))).toMatchSnapshot();
  });

  it('renders properly', async () => {
    expect(wrapper.find('PriceTag').length).toBe(1);
    expect(wrapper.find('PriceTag').text()).toBe('$35$30');
    expect(wrapper.find('DeleteProduct').length).toBe(1);
    expect(wrapper.find('DeleteProduct').props().id).toBe('pr0duct1d');
    expect(wrapper.find('DeleteProduct').props().userName).toBe('Miss Coleman Berge');
    expect(wrapper.find('DeleteProduct').text()).toBe('Delete');
    expect(wrapper.find('i.prdct-itm-actns').length).toBe(1);
    expect(wrapper.find('i.prdct-itm-actns').text()).toBe('(Offline)');
  });
});
