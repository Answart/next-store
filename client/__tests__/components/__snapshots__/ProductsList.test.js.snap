// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsList /> matches the snap shot 1`] = `
<ProductsList
  editView={true}
  products={
    Array [
      Object {
        "__typename": "Product",
        "brand": "Peggs",
        "category": "jewelry",
        "department": "accessories",
        "description": "Limited Edition watch from the 2018 Fall fashion line.",
        "id": "pr0duct1d",
        "image": Object {
          "__typename": "Image",
          "cloudinary_id": "111111",
          "height": 22,
          "id": "1m4g31d",
          "image_url": "peggswatch1.jpg",
          "large_image_url": "peggswatch2.jpg",
          "name": "peggswatch1",
          "transformation": "",
          "width": 22,
        },
        "online": false,
        "title": "Peggs Gold Edition Analog Watch Peggs Gold Edition Analog ",
        "user": Object {
          "__typename": "User",
          "id": "us3r1d",
          "name": "Miss Coleman Berge",
        },
        "variants": Array [
          Object {
            "__typename": "Variant",
            "availability": "4 in Stock!",
            "color": "white",
            "id": "v4r13nt1d",
            "image": Object {
              "__typename": "Image",
              "cloudinary_id": "111111",
              "height": 22,
              "id": "1m4g31d",
              "image_url": "peggswatch1.jpg",
              "large_image_url": "peggswatch2.jpg",
              "name": "peggswatch1",
              "transformation": "",
              "width": 22,
            },
            "price": 35,
            "product": Object {
              "__typename": "Product",
              "id": "pr0duct1d",
              "image": Object {
                "__typename": "Image",
                "cloudinary_id": "111111",
                "height": 22,
                "id": "1m4g31d",
                "image_url": "peggswatch1.jpg",
                "large_image_url": "peggswatch2.jpg",
                "name": "peggswatch1",
                "transformation": "",
                "width": 22,
              },
            },
            "quantity": 4,
            "sale": true,
            "salePrice": 30,
            "size": "S",
          },
        ],
      },
    ]
  }
  userId="us3r1d"
>
  <ProductStyles__StyledProductsList>
    <div
      className="ProductStyles__StyledProductsList-sc-2231u4-2 hbTbkM"
    >
      <ProductStyles__StyledProductsListItem
        key="pr0duct1d"
      >
        <div
          className="ProductStyles__StyledProductsListItem-sc-2231u4-3 eAuhwh"
        >
          <Link
            href={
              Object {
                "pathname": "/buy",
                "query": Object {
                  "id": "pr0duct1d",
                },
              }
            }
          >
            <a
              href="/buy?id=pr0duct1d"
              onClick={[Function]}
            >
              <img
                alt="Peggs Gold Edition Analog Watch Peggs Gold Edition Analog "
                height="610"
                src="peggswatch1.jpg"
                width="400"
              />
            </a>
          </Link>
          <div
            className="prdct-itm-info"
          >
            <Link
              href={
                Object {
                  "pathname": "/buy",
                  "query": Object {
                    "id": "pr0duct1d",
                  },
                }
              }
            >
              <a
                className="prdct-itm-title"
                href="/buy?id=pr0duct1d"
                onClick={[Function]}
              >
                Peggs Gold Edition Analog Watch Peggs Gold Edition Analog 
              </a>
            </Link>
            <div
              className="prdct-itm-price"
            >
              <PriceTag
                price={35}
                sale={true}
                salePrice={30}
              >
                <PriceTag__StyledPriceTag>
                  <div
                    className="PriceTag__StyledPriceTag-sc-119rsfa-0 ePOheW"
                  >
                    <span
                      className="line-through"
                    >
                      $35
                    </span>
                    <span
                      className="price-tag-padding price-tag-priceSale"
                    >
                      $30
                    </span>
                  </div>
                </PriceTag__StyledPriceTag>
              </PriceTag>
            </div>
          </div>
          <span
            className="prdct-itm-actns"
          >
            <Link
              href={
                Object {
                  "pathname": "/product/edit",
                  "query": Object {
                    "id": "pr0duct1d",
                  },
                }
              }
            >
              <a
                href="/product/edit?id=pr0duct1d"
                onClick={[Function]}
              >
                Edit
              </a>
            </Link>
            <DeleteProduct
              className="undrln-btn"
              id="pr0duct1d"
              userName="Miss Coleman Berge"
            >
              <Mutation
                mutation={
                  Object {
                    "definitions": Array [
                      Object {
                        "directives": Array [],
                        "kind": "OperationDefinition",
                        "name": Object {
                          "kind": "Name",
                          "value": "DELETE_PRODUCT_MUTATION",
                        },
                        "operation": "mutation",
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [
                                Object {
                                  "kind": "Argument",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "id",
                                  },
                                  "value": Object {
                                    "kind": "Variable",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                  },
                                },
                              ],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "deleteProduct",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                ],
                              },
                            },
                          ],
                        },
                        "variableDefinitions": Array [
                          Object {
                            "defaultValue": undefined,
                            "directives": Array [],
                            "kind": "VariableDefinition",
                            "type": Object {
                              "kind": "NonNullType",
                              "type": Object {
                                "kind": "NamedType",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "ID",
                                },
                              },
                            },
                            "variable": Object {
                              "kind": "Variable",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                            },
                          },
                        ],
                      },
                    ],
                    "kind": "Document",
                    "loc": Object {
                      "end": 96,
                      "start": 0,
                    },
                  }
                }
                update={[Function]}
                variables={
                  Object {
                    "id": "pr0duct1d",
                  }
                }
              >
                <button
                  className="undrln-btn"
                  disabled={false}
                  onClick={[Function]}
                >
                  Delete
                </button>
              </Mutation>
            </DeleteProduct>
          </span>
          <span>
            <i
              className="prdct-itm-actns"
            >
              (Offline)
            </i>
          </span>
        </div>
      </ProductStyles__StyledProductsListItem>
    </div>
  </ProductStyles__StyledProductsList>
</ProductsList>
`;
